## Genero los audios con ruido

#Audios para utilizar con la red y con el filtrado de wavelet.

La idea es generar audios para entrenar la red. En principio, necesitaría que la cantidad de audios sea par para poder dividir la red en 4:

* Audios con 0 dB de SNR
* Audios con -5 dB de SNR
* Audios con -10 dB de SNR
* Audios con -15 dB de SNR

Primero veo cuántos audios tengo...

In [1]:
import numpy as np
import os
import glob
import shutil

In [2]:
os.chdir(r'C:\Users\maxia\Desktop\Facultad\Labo de acústica\Codigo\4_audios_train')

audios = []

for dir in glob.iglob('*common*', recursive=True):
    audios.append(dir)

print(len(audios))
print(len(audios)/4)

22000
5500.0


In [3]:
#Cada audio tiene un total de muestras de:
ejemplo = np.load(audios[22])
print(f'Este audio tiene un total de {len(ejemplo)} muestras')

Este audio tiene un total de 65280 muestras


In [5]:
"""Con lo cual, voy a tener 3 grupos de 2087 y 1 grupo con 2090 audios """

'Con lo cual, voy a tener 3 grupos de 2087 y 1 grupo con 2090 audios '

# Generación de los audios ruidosos para entrenar:

Ya contamos con audios separados en largos de a 65280 muestras, ahora necesito agregarles ruido tal que su SNR sea de 0, 5, 10 o 15 dB

In [4]:
os.chdir(r'C:\Users\maxia\Desktop\Facultad\Labo de acústica\Codigo\4_audios_train')

print(len(audios))
print(audios[-1])

22000
3-common_voice_es_19976189_2.npy


In [5]:
#Separo los valores según cómo los voy a formar
ruido_0 = audios[0:5500]
ruido_5 = audios[5500:11000]
ruido_10 = audios[11000:16500]
ruido_15 = audios[16500:22000]

In [6]:
#Creo una función para agregar ruido para un determinado SNR

def noise(data, snr):

    rms = np.sqrt(np.mean(data**2)) #Calculo el rms de la señal
    STD_noise = np.sqrt((rms**2)/(10**(snr/10))) #Calculo el rms=std del ruido para x SNR
    noise = np.random.normal(0, STD_noise, data.shape) #Genero el ruido
    noisy_sigal = data + noise #Sumo señal y ruido

    return noisy_sigal

In [7]:
#Prueba con un audio random para ver si funciona
import sounddevice as sd
import time
os.chdir(r'C:\Users\maxia\Desktop\Facultad\Labo de acústica\Codigo\4_audios_train')

data = np.load(audios[0])

noise_data_0 = noise(data, 0)
noise_data_5 = noise(data, 5)
noise_data_10 = noise(data, 10)
noise_data_15 = noise(data, 15)

print('Audio limpio')
sd.play(data, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 0 dB SNR')
sd.play(noise_data_0, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 5 dB SNR')
sd.play(noise_data_5, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 10 dB SNR')
sd.play(noise_data_10, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 15 dB SNR')
sd.play(noise_data_15, samplerate=22050)

Audio limpio
Aduio + ruido a 0 dB SNR
Aduio + ruido a 5 dB SNR
Aduio + ruido a 10 dB SNR
Aduio + ruido a 15 dB SNR


In [8]:
print(audios[0].split('.npy')[0])

0-common_voice_es_18307761_1


In [9]:
#Genero los audios con ruido: para esto voy a agregar al principio de cada audio un número que represente su nivel de SNR
os.chdir(r'C:\Users\maxia\Desktop\Facultad\Labo de acústica\Codigo')

print('Creando audios a con SNR de 0 dB')
i=0
for filename in ruido_0:
    i+=1
    data = np.load('4_audios_train/'+filename)
    noise_data_0 = noise(data, 0)
    np.save('6_audios_ruidosos_train/'+filename.split('.npy')[0]+'_SNR-0', noise_data_0)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 0 dB de SNR')

    else:
        pass

print('Creando audios a 5 de SNR')
i=0
for filename in ruido_5:
    i+=1
    data = np.load('4_audios_train/'+filename)
    noise_data_5 = noise(data, 5)
    np.save('6_audios_ruidosos_train/'+filename.split('.npy')[0]+'_SNR-5', noise_data_5)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 5 de SNR')

    else:
        pass

print('Creando audios a 10 de SNR')
i=0
for filename in ruido_10:
    i+=1
    data = np.load('4_audios_train/'+filename)
    noise_data_10 = noise(data, 10)
    np.save('6_audios_ruidosos_train/'+filename.split('.npy')[0]+'_SNR-10', noise_data_10)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 10 de SNR')

    else:
        pass

print('Creando audios a 15 de SNR')
i=0
for filename in ruido_15:
    i+=1
    data = np.load('4_audios_train/'+filename)
    noise_data_15 = noise(data, 15)
    np.save('6_audios_ruidosos_train/'+filename.split('.npy')[0]+'_SNR-15', noise_data_15)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 15 de SNR')

    else:
        pass

print('Listo, audios guardados en la carpeta audios_ruidosos')

Creando audios a con SNR de 0 dB
Se crearon 100 audios a 0 dB de SNR
Se crearon 200 audios a 0 dB de SNR
Se crearon 300 audios a 0 dB de SNR
Se crearon 400 audios a 0 dB de SNR
Se crearon 500 audios a 0 dB de SNR
Se crearon 600 audios a 0 dB de SNR
Se crearon 700 audios a 0 dB de SNR
Se crearon 800 audios a 0 dB de SNR
Se crearon 900 audios a 0 dB de SNR
Se crearon 1000 audios a 0 dB de SNR
Se crearon 1100 audios a 0 dB de SNR
Se crearon 1200 audios a 0 dB de SNR
Se crearon 1300 audios a 0 dB de SNR
Se crearon 1400 audios a 0 dB de SNR
Se crearon 1500 audios a 0 dB de SNR
Se crearon 1600 audios a 0 dB de SNR
Se crearon 1700 audios a 0 dB de SNR
Se crearon 1800 audios a 0 dB de SNR
Se crearon 1900 audios a 0 dB de SNR
Se crearon 2000 audios a 0 dB de SNR
Se crearon 2100 audios a 0 dB de SNR
Se crearon 2200 audios a 0 dB de SNR
Se crearon 2300 audios a 0 dB de SNR
Se crearon 2400 audios a 0 dB de SNR
Se crearon 2500 audios a 0 dB de SNR
Se crearon 2600 audios a 0 dB de SNR
Se crearon 270

# Generación de los audios ruidosos para testing

In [11]:
os.chdir(r'C:\Users\maxia\Desktop\Facultad\Labo de acústica\Codigo\3_audios_test')

audios = []

for dir in glob.iglob('common*', recursive=True):
    audios.append(dir)

print(len(audios))
print(len(audios)/4)

1048
262.0


In [12]:
"""Con lo cual, voy a tener 4 grupos de 262 audios """

'Con lo cual, voy a tener 4 grupos de 262 audios '

In [13]:
#Separo los valores según cómo los voy a formar
ruido_0 = audios[0:262]
ruido_5 = audios[262:524]
ruido_10 = audios[524:786]
ruido_15 = audios[786:1048]

In [14]:
#Prueba con audios random
data = np.load(audios[0])

noise_data_0 = noise(data, 0)
noise_data_5 = noise(data, 5)
noise_data_10 = noise(data, 10)
noise_data_15 = noise(data, 15)

print('Audio limpio')
sd.play(data, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 0 dB SNR')
sd.play(noise_data_0, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 5 dB SNR')
sd.play(noise_data_5, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 10 dB SNR')
sd.play(noise_data_10, samplerate=22050)
time.sleep(len(data)/22050)

print('Aduio + ruido a 15 dB SNR')
sd.play(noise_data_15, samplerate=22050)

Audio limpio
Aduio + ruido a 0 dB SNR
Aduio + ruido a 5 dB SNR
Aduio + ruido a 10 dB SNR
Aduio + ruido a 15 dB SNR


In [15]:
print(audios[0].split('.npy')[0])

common_voice_es_18309707


In [16]:
#Genero los audios con ruido: para esto voy a agregar al principio de cada audio un número que represente su nivel de SNR
os.chdir(r'C:\Users\maxia\Desktop\Facultad\Labo de acústica\Codigo')

print('Creando audios a con SNR de 0 dB')
i=0
for filename in ruido_0:
    i+=1
    data = np.load('3_audios_test/'+filename)
    noise_data_0 = noise(data, 0)
    np.save('5_audios_ruidosos_test/'+filename.split('.npy')[0]+'_SNR-0', noise_data_0)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 0 dB de SNR')

    else:
        pass

print('Creando audios a 5 de SNR')
i=0
for filename in ruido_5:
    i+=1
    data = np.load('3_audios_test/'+filename)
    noise_data_5 = noise(data, 5)
    np.save('5_audios_ruidosos_test/'+filename.split('.npy')[0]+'_SNR-5', noise_data_5)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 5 de SNR')

    else:
        pass

print('Creando audios a 10 de SNR')
i=0
for filename in ruido_10:
    i+=1
    data = np.load('3_audios_test/'+filename)
    noise_data_10 = noise(data, 10)
    np.save('5_audios_ruidosos_test/'+filename.split('.npy')[0]+'_SNR-10', noise_data_10)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 10 de SNR')

    else:
        pass

print('Creando audios a 15 de SNR')
i=0
for filename in ruido_15:
    i+=1
    data = np.load('3_audios_test/'+filename)
    noise_data_15 = noise(data, 15)
    np.save('5_audios_ruidosos_test/'+filename.split('.npy')[0]+'_SNR-15', noise_data_15)

    if(i%100) == 0:
        print(f'Se crearon {i} audios a 15 de SNR')

    else:
        pass

print('Listo, audios guardados en la carpeta audios_ruidosos')

Creando audios a con SNR de 0 dB
Se crearon 100 audios a 0 dB de SNR
Se crearon 200 audios a 0 dB de SNR
Creando audios a 5 de SNR
Se crearon 100 audios a 5 de SNR
Se crearon 200 audios a 5 de SNR
Creando audios a 10 de SNR
Se crearon 100 audios a 10 de SNR
Se crearon 200 audios a 10 de SNR
Creando audios a 15 de SNR
Se crearon 100 audios a 15 de SNR
Se crearon 200 audios a 15 de SNR
Listo, audios guardados en la carpeta audios_ruidosos
